### ISSUE TITLE
Feature request: instance norm without the computation of running statistics

### ISSUE THREAD

**Issue by netheril96** [pytorch/pytorch#4509]
The current instance norm implementation delegates to `nn.functional.batch_norm` for the actual computation, and if my understanding is not wrong, the latter requires the presence of running mean and variance. Instance norm, however, usually never needs these statistics even in testing mode, so it is waste of both time and space. It would be nice if `batch_norm` could be modified such that running mean and variance are optional. It won't be hard, will it?

**Comment by apaszke** [comment]
Yeah that should be fairly easy

**Comment by netheril96** [comment]
@apaszke Could you do it then? I am not familiar enough with PyTorch sources to propose a PR, especially since the implementation of `nn.functional` is very hard to understand.

**Comment by apaszke** [comment]
I'm ok with accepting a PR, but I don't think it's a priority for us at this point, so I'll continue working on other things. Of course if you show us an example where getting rid of running stats makes the training much faster then we will reconsider the priority.

**Comment by netheril96** [comment]
😅It will be a little hard to benchmark the exact difference of memory and time usage, without an implementation that does not require the running statistics.

**Comment by ngimel** [comment]
Then show what fraction of runtime instance norm as a whole is taking? If it's 10%, improving it by 20% is not going to give big overall speed-up.

**Comment by Kaixhin** [comment]
For context there are two PRs on LayerNorm (https://github.com/pytorch/pytorch/pull/2019 and https://github.com/pytorch/pytorch/pull/2112) with some discussion on this (one interpretation of InstanceNorm is that it is LayerNorm generalised to > 1D).

@ngimel Both LayerNorm and InstanceNorm are normalisation functions that do not need context -
 BatchNorm does need to keep running stats because of the way it works though. Hence one could make a purely functional form of the former without the need to specify the size of the inputs in advance. In practice, most people will use fixed-size inputs and using optimised BatchNorm code is probably faster. Any further thoughts?

**Comment by netheril96** [comment]
I am not familiar with the C sources of PyTorch, so I have a naive question: could the fused kernel of batch norm be more generalized? Not just by making running mean and variance optional, but also by generalizing the dimensions to normalize over. Instance norm and layer norm can then be treated as first class citizens, just like batch norm.

**Comment by ngimel** [comment]
@Kaixhin agree on not needing running stats on layer norm/instance norm, however, for both of them you still need trainable scale and shift, right? Don't the dimensions of those depend on the input dimensions?
@netheril96, right now pytorch in most cases uses cudnn implementation of batch norm, rather than THCUNN. IIRC, cudnn does not have to calculate running statistics (i.e. it will work if one passes null pointers for running_var and running_mean), but pytorch bindings will have to be changed to make use of that. Generalizing dimensions to normalize over is a different story, to a degree it can be worked around by having views on tensor, but it does not cover all possible cases.

**Comment by Kaixhin** [comment]
@ngimel they do not necessarily need trainable shifts and scales, but in that case yes size needs to be known in advance. I believe that most of the time people will use trainable shifts and scales, so it may be a lower priority in that sense.

**Comment by Mypathissional** [comment]
Hi, 
I am a little confused on InstanceNormalization with tracking_running_estimates=True. Is it any different from the usual BatchNormalizationLayer? 
Would be grateful for any clarifications.

**Comment by cdeln** [comment]
Bumping this issue. I believe the implementation of instance normalization in evaluation mode is incorrect.
Instance normalization (according to both the original paper and the PyTorch documentation) is applied for each batch sample and channel independently (i.e. it only normalizes over spatial dimensions).
All normalization variables are computed "online" for each batch, there is nothing to "record" from training (i.e. by setting `track_running_stats=True`.

Background to why I am bumping this: I am exporting a PyTorch model to ONNX and I am given a warning about instance normalization layer being in training mode despite putting the model in evaluation mode. Result is that inference results differ from results during training.

@ngimel Indeed, the learnable mean and variance depend on the number of channels, but only that dimension. 

@ngimel  Is passing `running_var` and `running_mean` null pointers implemented in the core code yet? I do not see how this layer can function in evaluation mode otherwise.

@soumith Can you elaborate on why #4922 closes this issue?
